janus_server/src/binary_utils/job_driver.rs

```diff
@@ -0,0 +1,414 @@
+        min_job_discovery_delay: Duration,
+        max_job_discovery_delay: Duration,
+        max_concurrent_job_workers: usize,
+        worker_lease_duration: Duration,
+        worker_lease_clock_skew_allowance: Duration,
+        incomplete_job_acquirer: JobAcquirer,
+        job_stepper: JobStepper,
+        job_stepper_context: JobStepperContext,
+    ) -> Self {
+        Self {
+            datastore,
+            clock,
+            min_job_discovery_delay,
+            max_job_discovery_delay,
+            max_concurrent_job_workers,
+            worker_lease_duration,
+            worker_lease_clock_skew_allowance,
+            incomplete_job_acquirer,
+            job_stepper,
+            job_stepper_context,
+        }
+    }
+
+    /// Run this job driver, periodically seeking incomplete jobs and stepping them.
+    #[tracing::instrument(skip(self))]
+    pub async fn run(self: Arc<Self>) -> ! {
+        let sem = Arc::new(Semaphore::new(self.max_concurrent_job_workers));
+        let mut job_discovery_delay = Duration::ZERO;
+
+        loop {
+            // Wait out our job discovery delay, if any.
+            time::sleep(time::Duration::from_secs(job_discovery_delay.as_seconds())).await;
+
+            // Wait until we are able to start at least one worker. (permit will be immediately released)
+            //
+            // Unwrap safety: Semaphore::acquire is documented as only returning an error if the
+            // semaphore is closed, and we never close this semaphore.
+            let _ = sem.acquire().await.unwrap();
+
+            // Acquire some jobs which are ready to be stepped.
+            //
+            // We determine the maximum number of jobs to acquire based on the number of semaphore
+            // permits available, since we'd like to start processing any acquired jobs immediately
+            // to avoid potentially timing out while waiting on _other_ jobs to finish being
+            // stepped. This is racy given that workers may complete (and relinquish their permits)
+            // concurrently with us acquiring jobs; but that's OK, since this can only make us
+            // underestimate the number of jobs we can acquire, and underestimation is acceptable
+            // (we'll pick up any additional jobs on the next iteration of this loop). We can't
+            // overestimate since this task is the only place that permits are acquired.
+
+            // TODO(brandon): only acquire jobs whose batch units have not already been collected (probably by modifying acquire_incomplete_aggregation_jobs)
+            let max_acquire_count = sem.available_permits();
+            info!(max_acquire_count, "Acquiring jobs");
+            let acquired_jobs = (self.incomplete_job_acquirer)(
+                self.datastore.clone(),
+                self.worker_lease_duration,
+                max_acquire_count,
+            )
+            .await;
+            let acquired_jobs = match acquired_jobs {
```

Comment:
I probably should have done this myself in the agg job driver, but especially now that the "acquire" process has been made generic, IMO we should have a check that the number of acquired jobs is <= `max_acquire_count`. This is required for some of the reasoning in the comments around semaphore acquiry to work out. An assert would be fine IMO since this can be viewed as programmer error.